src/main/java/org/apache/sysml/runtime/controlprogram/paramserv/LocalParamServer.java

```diff
@@ -24,36 +24,37 @@
 import org.apache.sysml.runtime.controlprogram.context.ExecutionContext;
 import org.apache.sysml.runtime.instructions.cp.Data;
 import org.apache.sysml.runtime.instructions.cp.ListObject;
+import org.apache.sysml.runtime.instructions.cp.ParamservBuiltinCPInstruction;
 
 public class LocalParamServer extends ParamServer {
 
 	public LocalParamServer(ListObject model, String aggFunc, Statement.PSFrequency freq,
 			Statement.PSUpdateType updateType, ExecutionContext ec, int workerNum,
-			ListObject hyperParams) {
-		super(model, aggFunc, freq, updateType, ec, workerNum, hyperParams);
+			ParamservBuiltinCPInstruction.PSErrorHandler handler) {
+		super(model, aggFunc, freq, updateType, ec, workerNum, handler);
 	}
 
 	@Override
 	public void push(long workerID, ListObject gradients) {
-		synchronized (_lock) {
-			_queue.add(new Gradient(workerID, gradients));
-			_lock.notifyAll();
+		try {
+			_gradientsQueue.put(new Gradient(workerID, gradients));
+			launchService();
```

Comment:
It seems that there is actually no need to run the aggregation service as a separate thread. The single-threaded thread pool provides you serialization of incoming gradients. You might be able to simple call the code of the run method as a synchronized method (serialized on the parameter server instance).